src/realm/alloc.hpp

```diff
@@ -297,14 +297,21 @@ inline int_fast64_t from_ref(ref_type v) noexcept
 {
     // Check that v is divisible by 8 (64-bit aligned).
     REALM_ASSERT_DEBUG(v % 8 == 0);
-    return util::from_twos_compl<int_fast64_t>(v);
+
+    // Unfortunately we have to rely on this to be bit pattern preserving on
+    // our platforms due to the fact that we still have places where we do
+    // not use these functions.
+    return int_fast64_t(v);
 }
```

Comment:
The following form is superior because it avoid undefined behavior when `v` has its 63rd bit set:

    return util::from_twos_compl<int_fast64_t>(uint_fast64_t(v));

On contemporary arcs, it reduces to the same as `return int_fast64_t(v);`.